### ISSUE TITLE
WeakKeyDict regression (`get!`)

### ISSUE THREAD

**Issue by yuyichao** [JuliaLang/julia#20383]
`get!(::WeakKeyDict)` appears to have regressed in the past 55 days on master.

On a 55 days master and on 0.5.

```julia
julia> d = WeakKeyDict{Task,Int}()
WeakKeyDict{Task,Int64} with 0 entries

julia> get!(d, current_task(), 1)
1
```

Showing the non-empty `d` errors on the old versions. The show/iteration is fixed on master but `get!` is now broken.

```julia
julia> d = WeakKeyDict{Task,Int}()
WeakKeyDict{Task,Int64} with 0 entries

julia> get!(d, current_task(), 1)
ERROR: MethodError: Cannot `convert` an object of type Task to an object of type WeakRef
This may have arisen from a call to the constructor WeakRef(...),
since type constructors fall back to convert methods.
Stacktrace:
 [1] get!(::Function, ::Dict{WeakRef,Int64}, ::Task) at ./dict.jl:436
 [2] lock(::Base.##306#307{WeakKeyDict{Task,Int64},Task,Int64}, ::Base.Threads.RecursiveTatasLock) at ./lock.jl:101
 [3] get!(::WeakKeyDict{Task,Int64}, ::Task, ::Int64) at ./weakkeydict.jl:107
```

**Comment by vtjnash** [comment]
It looks like the implementation of `get!` is wrong for WeakKeyDict, since it bypasses our `setindex!`, it also is bypassing the WeakRef logic for creating keys. I don't think this was ever quite correct.

**Comment by JeffBezanson** [comment]
Fixed by #28198
